sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs

```diff
@@ -0,0 +1,129 @@
+    {
+        Argument.AssertNotNull(credential, nameof(credential));
+        Argument.AssertNotNullOrEmpty(headerName, nameof(headerName));
+
+        return new ApiKeyAuthenticationPolicy(credential, headerName, KeyLocation.Header, keyPrefix);
+    }
+
+    public static ApiKeyAuthenticationPolicy CreateQueryApiKeyPolicy(ApiKeyCredential credential, string queryName)
+    {
+        // TODO: Add tests for this implementation if the API is approved.
+        Argument.AssertNotNull(credential, nameof(credential));
+        Argument.AssertNotNullOrEmpty(queryName, nameof(queryName));
+
+        return new ApiKeyAuthenticationPolicy(credential, queryName, KeyLocation.Query);
+    }
+
+    public static ApiKeyAuthenticationPolicy CreateBasicAuthorizationPolicy(ApiKeyCredential credential)
+    {
+        Argument.AssertNotNull(credential, nameof(credential));
+
+        return new ApiKeyAuthenticationPolicy(credential, "Authorization", KeyLocation.Header, "Basic");
+    }
+
+    public static ApiKeyAuthenticationPolicy CreateBearerAuthorizationPolicy(ApiKeyCredential credential)
+    {
+        Argument.AssertNotNull(credential, nameof(credential));
+
+        return new ApiKeyAuthenticationPolicy(credential, "Authorization", KeyLocation.Header, "Bearer");
+    }
+
+    private ApiKeyAuthenticationPolicy(ApiKeyCredential credential, string name, KeyLocation keyLocation, string? keyPrefix = null)
+    {
+        Argument.AssertNotNull(credential, nameof(credential));
+        Argument.AssertNotNullOrEmpty(name, nameof(name));
+
+        _credential = credential;
+
+        _name = name;
+        _location = keyLocation;
+        _keyPrefix = keyPrefix;
+    }
+
+    public sealed override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        SetKey(message);
+
+        ProcessNext(message, pipeline, currentIndex);
+    }
+
+    public sealed override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        SetKey(message);
+
+        await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+    }
+
+    private void SetKey(PipelineMessage message)
+    {
+        switch (_location)
+        {
+            case KeyLocation.Header:
+                SetHeader(message);
+                break;
+            case KeyLocation.Query:
+                AddQueryParameter(message);
+                break;
+            default:
+                throw new InvalidOperationException($"Unsupported value for Key location: '{_location}'.");
+        }
+    }
+
+    private void SetHeader(PipelineMessage message)
+    {
+        _credential.Deconstruct(out string key);
+
+        message.Request.Headers.Set(_name, _keyPrefix != null ? $"{_keyPrefix} {key}" : key);
+    }
+
+    private void AddQueryParameter(PipelineMessage message)
+    {
+        _credential.Deconstruct(out string key);
```

Comment:
It's unclear what the correct logic is here with regard to escaping the key value and efficiently implementing idempotency for retries ... I'll look into whether we have an implementation of this policy in Azure.Core today, but if we don't, we can add this in a later version per expediency.

Reply:
Since this is a static key-based policy, I think the correct way to handle retry idempotency would be to only add this policy as a per-call rather than pre-retry policy.

Reply:
@KrzysztofCwalina had suggested a design that enforces the pipeline position for policies - I don't have an issue tracking it yet but will file one.

Reply:
Tracking with https://github.com/Azure/azure-sdk-for-net/issues/41392

Reply:
That guidance makes sense, but I'd call the key rotation on retry an edge case. Most intentional key rotations are done with overlap between the validity of the old and new keys. Otherwise you have a race between rotation and key update to avoid an outage.